airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/streams.py

```diff
@@ -0,0 +1,448 @@
+
+        # add the 'query' parameter
+        res = self._prepare_query(updated_after)
+        res.update(
+            {
+                "sort_by": "created_at",
+                "sort_order": "asc",
+                "size": self.state_checkpoint_interval,
+            }
+        )
+        if next_page_token:
+            res["page"] = next_page_token["next_page"]
+        return res
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        # try to save maximum value of a cursor field
+
+        return {
+            self.cursor_field: max(
+                (latest_record or {}).get(self.cursor_field, ""), (current_stream_state or {}).get(self.cursor_field, "")
+            )
+        }
+
+
+class IncrementalBasicEntityStream(IncrementalBasicSearchStream, ABC):
+    """basic stream for endpoints where an entity name can be used in a path value
+    https://<subdomain>.zendesk.com/api/v2/<entity_name>.json
+    """
+
+    # for generation of a path value and as rule as JSON root name of all response
+    entity_type: str = None
+
+    # for partial cases when JSON root name of responses is not equal a entity_type value
+    response_list_name: str = None
+
+    def path(self, *args, **kwargs) -> str:
+        return f"{self.entity_type}.json"
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """returns data from API AS IS"""
+        yield from response.json()[self.response_list_name or self.entity_type] or []
+
+
+class IncrementalBasicUnsortedStream(IncrementalBasicEntityStream, ABC):
+    """basic stream for loading without sorting
+
+    Some endpoints don't provide approachs for data filtration
+    We can load all reconds fully and select updated data only
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # For saving of a last stream value. Not all functions provides this value
+        self._cursor_date = None
+        # Flag for marking of completed process
+        self._finished = False
+        # For saving of a relevant last updated date
+        self._max_cursor_date = None
+        # For changing of filter logic
+        self._updated_cursor_field = None
+
+    def _save_cursor_state(self, state: Mapping[str, Any] = None):
+        """need to save stream state for some internal logic"""
+        if not self._cursor_date and state and state.get(self.cursor_field):
+            self._cursor_date = self.str2datetime(state[self.cursor_field])
+        return
+
+    def request_params(
+        self, stream_state: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None, **kwargs
+    ) -> MutableMapping[str, Any]:
+        self._save_cursor_state(stream_state)
+        return {}
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
```

Comment:
it looks quite complicated for me. Maybe it is only for me. Lets other team members review this part as well. Please ask Bazarnov and Dudar to review this as well.